sources/api/thar-be-updates/src/main.rs

```diff
@@ -0,0 +1,562 @@
+    }
+    Ok(false)
+}
+
+/// Queries the Bottlerocket API and returns the 'ignore-waves' setting. If it doesn't exist for whatever reason, return false
+fn get_ignore_waves(socket_path: &str) -> Result<bool> {
+    let uri = "/settings";
+    let method = "GET";
+    let (code, response_body) = apiclient::raw_request(&socket_path, uri, method, None)
+        .context(error::APIRequest { method, uri })?;
+    ensure!(
+        code.is_success(),
+        error::APIResponse {
+            method,
+            uri,
+            code,
+            response_body,
+        }
+    );
+    let settings: serde_json::Value =
+        serde_json::from_str(&response_body).context(error::ResponseJson { uri })?;
+    Ok(settings["updates"]["ignore-waves"]
+        .as_bool()
+        .unwrap_or(false))
+}
+
+/// Prepares the update by downloading and writing the update to the staging partition
+fn prepare_update(
+    status: &mut UpdateStatus<update_metadata::Update>,
+    ignore_waves: bool,
+) -> Result<()> {
+    debug!("Spawning 'updog update-image'");
+    let chosen_update = status
+        .get_chosen_update()
+        .context(error::ChosenUpdate)?
+        .clone();
+    // Prepare arguments, append '--now' if we're ignoring waves
+    let chosen_version = chosen_update.get_version().to_string();
+    let mut args = vec!["update-image", "--image", &chosen_version];
+    if ignore_waves {
+        args.push("--now");
+    }
+    let output = Command::new("updog")
+        .args(args)
+        .output()
+        .context(error::Updog)?;
+    status.set_recent_command_info(UpdateCommand::PrepareUpdate, Some(&output), None);
+    if output.status.success() {
+        status.set_staging_partition_image_info(chosen_update);
+    } else {
+        warn!("Failed to prepare the update with updog");
+    }
+    Ok(())
+}
+
+/// "Activates" the staged update by setting the appropriate priority flags in the partition table
+fn activate_update(status: &mut UpdateStatus<update_metadata::Update>) -> Result<()> {
+    let mut gpt_state = State::load().context(error::PartitionTableRead)?;
+    gpt_state
+        .upgrade_to_inactive()
+        .context(error::InactivePartitionUpgrade)?;
+    gpt_state.write().context(error::PartitionTableWrite)?;
+    status.mark_staging_partition_next_to_boot()?;
+    Ok(())
+}
+
+/// "Deactivates" the staged update by rolling back actions done by `activate_update`
+fn deactivate_update(status: &mut UpdateStatus<update_metadata::Update>) -> Result<()> {
+    let mut gpt_state = State::load().context(error::PartitionTableRead)?;
+    // This actually wipes all the priority bits in the inactive partition
+    gpt_state.cancel_upgrade();
+    // The update image is still there in the inactive partition so mark as valid
+    gpt_state.mark_inactive_valid();
+    gpt_state.write().context(error::PartitionTableWrite)?;
+    status.unmark_staging_partition_next_to_boot()?;
+    Ok(())
+}
```

Comment:
I'd much prefer doing this work in `updog`; I think it's too close to the implementation details.

Seems like `activate_update` should call `updog image-apply`, and `deactivate_update` should call `updog image-revert`.

Reply:
While `updog` has `image-apply`. `updog` doesn't have `image-revert` or any subcommand to revert `image-apply`. It would need to be done through `signpost`, either the binary, or the library calls like what's currently here.

Reply:
I think we should consider adding `image-revert` to updog rather than adding such low-level logic here.